src/workerd/api/node/i18n.c++

```diff
@@ -0,0 +1,258 @@
+  auto utf16_input =
+      kj::arrayPtr<char16_t>(reinterpret_cast<char16_t*>(source.begin()), source.size() / 2);
+  size_t expected_utf8_length =
+      simdutf::utf8_length_from_utf16le(utf16_input.begin(), utf16_input.size());
+
+  // Workers are limited to 128MB so this isn't actually a realistic concern, but sanity check.
+  JSG_REQUIRE(expected_utf8_length <= ISOLATE_LIMIT, Error,
+              "Expected UTF-8 length is too large to transcode");
+
+  auto destbuf = kj::heapArray<kj::byte>(expected_utf8_length);
+
+  size_t actual_length = simdutf::convert_utf16le_to_utf8(utf16_input.begin(), utf16_input.size(),
+                                                          destbuf.asChars().begin());
+  JSG_REQUIRE(actual_length == expected_utf8_length, Error, "Expected UTF8 length mismatch");
+
+  // simdutf returns 0 for invalid UTF-8 value.
+  if (actual_length == 0) {
+    return kj::none;
+  }
+
+  return destbuf.asBytes().attach(kj::mv(destbuf));
+}
+
+} // namespace
+
+Converter::Converter(Encoding encoding, kj::StringPtr substitude) {
+  UErrorCode status = U_ZERO_ERROR;
+  auto name = getEncodingName(encoding);
+  auto conv = ucnv_open(name, &status);
+  JSG_REQUIRE(U_SUCCESS(status), Error, "Failed to initialize converter");
+  conv_ = kj::Own<UConverter>(conv, ConverterDisposer::INSTANCE);
+  setSubstitudeChars(substitude);
+}
+
+Converter::Converter(UConverter* converter, kj::StringPtr substitude)
+    : conv_(converter, ConverterDisposer::INSTANCE) {
+  setSubstitudeChars(substitude);
+}
+
+UConverter* Converter::conv() const {
+  return const_cast<UConverter*>(conv_.get());
+}
+
+size_t Converter::maxCharSize() const {
+  KJ_ASSERT_NONNULL(conv_.get());
+  return ucnv_getMaxCharSize(conv_.get());
+}
+
+size_t Converter::minCharSize() const {
+  KJ_ASSERT_NONNULL(conv_.get());
+  return ucnv_getMinCharSize(conv_.get());
+}
+
+void Converter::reset() {
+  KJ_ASSERT_NONNULL(conv_.get());
+  ucnv_reset(conv_.get());
+}
+
+void Converter::setSubstitudeChars(kj::StringPtr sub) {
+  KJ_ASSERT_NONNULL(conv_.get());
+  UErrorCode status = U_ZERO_ERROR;
+  if (sub.size() > 0) {
+    ucnv_setSubstChars(conv_.get(), sub.begin(), sub.size(), &status);
+    JSG_REQUIRE(U_SUCCESS(status), Error, "Setting ICU substitude characters failed");
+  }
+}
+
+kj::Array<kj::byte> transcode(kj::ArrayPtr<kj::byte> source, Encoding fromEncoding,
+                              Encoding toEncoding) {
+  TranscodeImpl transcode_function = &TranscodeDefault;
+  switch (fromEncoding) {
+  case Encoding::ASCII:
+  case Encoding::LATIN1:
+    if (toEncoding == Encoding::UTF16LE) {
+      transcode_function = &TranscodeToUCS2;
+    }
+    break;
+  case Encoding::UTF8:
+    if (toEncoding == Encoding::UTF16LE) {
+      transcode_function = &TranscodeUcs2FromUtf8;
+    }
+    break;
+  case Encoding::UTF16LE:
+    switch (toEncoding) {
+    case Encoding::UTF16LE:
+      transcode_function = &TranscodeDefault;
```

Comment:
Can't we optimize the case of fromEncoding == toEncoding to just be return copy of src?

Reply:
Yeah, nice catch. Updating the implementation.